api/cas/account.py

```diff
@@ -0,0 +1,364 @@
+    :raises: ValidationError, PermissionDenied, APIException
+    """
+
+    # check required fields
+    email = data_user.get('email')
+    token = data_user.get('verificationCode')
+    if not email or not token:
+        raise ValidationError(detail=messages.INVALID_REQUEST)
+
+    # retrieve the user
+    user = util.find_user_by_email_or_guid(None, email, username_only=False)
+    if not user:
+        raise APIException(detail=messages.EMAIL_NOT_FOUND)
+
+    # check the token and its verification
+    if token not in user.email_verifications:
+        raise PermissionDenied(detail=messages.INVALID_CODE)
+    verification = user.email_verifications[token]
+    email = verification['email']
+    provider = verification['external_identity'].keys()[0]
+    identity = verification['external_identity'][provider].keys()[0]
+    if provider not in user.external_identity:
+        raise PermissionDenied(detail=messages.INVALID_CODE)
+
+    external_identity_status = user.external_identity[provider][identity]
+    util.ensure_external_identity_uniqueness(provider, identity, user)
+
+    # register/update user, set identity status to verified, clear pending verifications and send emails
+    if not user.is_registered:
+        user.register(email)
+    if not user.emails.filter(address=email.lower()):
+        user.emails.create(address=email.lower())
+    user.date_last_logged_in = timezone.now()
+    user.external_identity[provider][identity] = 'VERIFIED'
+    user.social[provider.lower()] = identity
+    del user.email_verifications[token]
+    if external_identity_status == 'CREATE':
+        send_mail(to_addr=user.username, mail=WELCOME, mimetype='html', user=user)
+    elif external_identity_status == 'LINK':
+        send_mail(to_addr=user.username, mail=EXTERNAL_LOGIN_LINK_SUCCESS, user=user, external_id_provider=provider)
+
+    # generate v1 key for CAS login
+    user.verification_key = generate_verification_key(verification_type=None)
+    user.save()
+
+    return user, external_identity_status
+
+
+def handle_password_forgot(data_user):
+    """
+    Find account by email and verify if the user is eligible for reset password. If so, send the verification email
+    if user hasn't recently make the same request.
+
+    :param data_user: the user
+    :return: the user with updated pending password reset verification
+    :raises: ValidationError, PermissionDenied, APIException
+    """
+
+    # check required fields
+    email = data_user.get('email')
+    if not email:
+        raise ValidationError(detail=messages.INVALID_REQUEST)
+
+    # retrieve the user
+    user = util.find_user_by_email_or_guid(None, email, username_only=False)
+    if not user:
+        raise PermissionDenied(detail=messages.EMAIL_NOT_FOUND)
+
+    # check throttle
+    if not throttle_period_expired(user.email_last_sent, web_settings.SEND_EMAIL_THROTTLE):
+        raise PermissionDenied(detail=messages.EMAIL_THROTTLE_ACTIVE)
```

Comment:
Update: CAS uses `org.apache.http.HttpStatus`  which does not define a static const for `429`. I decide to use the general `400` instead of extend the interface. Ref: https://hc.apache.org/httpcomponents-core-ga/httpcore/apidocs/org/apache/http/HttpStatus.html